### PR TITLE
feat(nested-errors,custom-errors): add nested errors for field array

### PR DESCRIPTION
**Pull request by thislogancall**
Add the ability for fields within a field array to have their own validation and errors.

Add the option to add a custom error message for each validation schema.

### CODE REVIEW

**Comment by fakenickels** [comment]
this is very cool! Will review it ASAP

**Comment by thislogancall** [comment]
If it's helpful, here's an example of how NestedErrors are being used:

Form init and Validation schema:
```reason
/** InviteUsersForm.re **/
open ReForm;
open ReSchema;

module AuthedCompanyInviteUsersLenses = [%lenses
  type state = {
    message: string,
    users: array(invitee),
  }
];

module Form = ReForm.Make(AuthedCompanyInviteUsersLenses);

let schema =
  Form.Validation.Schema([|
    Custom(
      Users,
      ({users}) =>
        if (users->Array.length < 1) {
          Error("Users are required.");
        } else {
          let errors =
            users->Belt.Array.reduceWithIndex([||], (errors, user, index) =>
              if (user.id == "") {
                Belt.Array.concat(
                  errors,
                  [|{name: "email", index, error: "Email is required"}|],
                );
              } else if (!Validate.email(user.id)) {
                Belt.Array.concat(
                  errors,
                  [|{name: "email", index, error: "Invalid Email"}|],
                );
              } else {
                errors;
              }
            );

          if (errors->Array.length > 0) {
            NestedErrors(errors);
          } else {
            Valid;
          };
        },
    ),
  |]);

/** FieldError.re **/
[@react.component]
let make = (~error, ~index=None, ~name=None) =>
  switch (error, index, name) {
  | (Ok, _, _) => React.null
  | (Errors(errors), Some(index), Some(name)) =>
    switch (
      errors
      |> Js.Array.find(error => error.index == index && error.name == name)
    ) {
    | None => React.null
    | Some(error) =>
      <Alert type_=`error className="error-container">
        error.error->React.string
      </Alert>
    }
  | (Error(error), _, _) =>
    <Alert type_=`error className="error-container">
      error->React.string
    </Alert>
  };

/** Usage of FormError (use this within loop of array fields to get accurate error)**/
<FieldError
    error={form.getFieldErrorState(Field(Users))}
    index={Some(index)}
    name={Some("userRoleName")}
/>
```

File: packages/bs-reform/src/ReForm.re [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
   let get: (state, field('a)) => 'a;
 };
+
 type fieldState =
   | Pristine
   | Valid
+  | NestedErrors(array(ReSchema.childFieldError))
   | Error(string);
+
+type fieldError =
```
Comment (fakenickels) [inline packages/bs-reform/src/ReForm.re]: hmm why not add to the `fieldState` Error constructor?

File: packages/bs-reform/src/ReForm.re [hunk 2]
```diff
@@ -93,22 +106,43 @@ module Make = (Config: Config) => {
       let Validation.Schema(validators) = schema;
       validators->Belt.Array.map(validator =>
         switch (validator) {
-        | Validation.IntMin(field, _min) => (
+        | Validation.IntMin(field, _min, _customErrorMessage) => (
```
Comment (fakenickels) [inline packages/bs-reform/src/ReForm.re]: hmm custom messages could be passed via `i18n` dictionary

File: packages/bs-reform/src/ReForm.re [hunk 3]
```diff
@@ -414,18 +468,52 @@ module Make = (Config: Config) => {
...
+      );
+
+    let getNestedFieldError = (field, index, name) =>
+      getFieldState(field)
+      |> (
+        fun
+        | NestedErrors(errors) => {
+            switch (
+              errors
+              |> Js.Array.find(error =>
```
Comment (fakenickels) [inline packages/bs-reform/src/ReForm.re]: if `errors` are sorted by index we could access via `Belt.Array.get`

File: packages/reschema/src/ReSchema.re [hunk 4]
```diff
@@ -5,30 +5,37 @@ module type Lenses = {
   let get: (state, field('a)) => 'a;
 };
 
+type childFieldError = {
```
Comment (fakenickels) [inline packages/reschema/src/ReSchema.re]: I like the naming here!

File: packages/reschema/src/ReSchema.re [hunk 5]
```diff
@@ -5,30 +5,37 @@ module type Lenses = {
...
+  error: string,
+  index: int,
+  name: string,
```
Comment (fakenickels) [inline packages/reschema/src/ReSchema.re]: hmm wonder if we could fit into the abstraction a `'a` kind of type to let the users pass the actual GADT constructor, still thinking here

File: packages/bs-reform/src/ReForm.re (hunk 1 above)
Comment (thislogancall) [reply packages/bs-reform/src/ReForm.re]: Good point. I think initially I was thinking it made sense to have it separate so users have simpler access to the error within the fieldInterface, but now I'm thinking it just adds bloat.

File: packages/bs-reform/src/ReForm.re (hunk 3 above)
Comment (thislogancall) [reply packages/bs-reform/src/ReForm.re]: Agreed.

File: packages/bs-reform/src/ReForm.re (hunk 2 above)
Comment (thislogancall) [reply packages/bs-reform/src/ReForm.re]: I've looked through ReSchema.re and ReSchemaI18n.re and am not sure how you would accomplish this. Can you give me an example of what you're thinking?

Comment (fakenickels) [reply packages/bs-reform/src/ReForm.re]: ofc! 
here's how:

```reason
let formErrorsDictionary: ReSchemaI18n.t = {
  ...ReSchemaI18n.default,
  stringMin: (~value as _value, ~min) =>
    "Este valor deve ter pelo menos " ++ string_of_int(min) ++ " caracteres",
  stringMax: (~value as _value, ~max) =>
    "Este valor deve ter no máximo " ++ string_of_int(max) ++ " caracteres",
};

[@react.component]
let make = (~onSubmit, ~navigation, ~initialState) => {
  module AppConfig = (val AppConfigTypes.useAppConfig());

  let snackbar = Snackbar.useSnackbar();

  let reform: ProfileFormHook.api =
    ProfileFormHook.use(
      ~i18n=formErrorsDictionary,
      ~schema=
        ProfileFormHook.Validation.Schema([|
```

Comment (thislogancall) [reply packages/bs-reform/src/ReForm.re]: My initial impression when seeing this is that it doesn't allow for validation error messages on a per-field basis. While you could do something more general for errors like "This field is required", I also think there is a strong use case to have errors custom to that specific field like "A username is required." This is also helpful if it's a large form and a summary of errors is needed at the bottom of the form.

I haven't tried this but maybe another option would be to have a second "Custom" validator that allows you to pass in a default validator and a custom error message?

What are your thoughts here?

Comment (thislogancall) [reply packages/bs-reform/src/ReForm.re]: @fakenickels Any thoughts here?

Comment (fakenickels) [reply packages/bs-reform/src/ReForm.re]: >I haven't tried this but maybe another option would be to have a second "Custom" validator that allows you to pass in a default validator and a custom error message?

I don't know if I could get what you are trying to say, but isn't that how `Custom` works already?

**Comment by fakenickels** [comment]
Sorry about the delay @thislogancall, will pick up the discussion now

File: packages/bs-reform/src/ReForm.re (hunk 2 above)
Comment (thislogancall) [reply packages/bs-reform/src/ReForm.re]: Right, so the difference is that Custom requires you put a custom validation and custom error message. What I'm suggesting is a way to use currently predefined validations but with custom error messages.

**Comment by fakenickels** [comment]
Picked up at #171

**Pull request closed without merging**